### PR TITLE
t2 star relaxometry estimation pipeline

### PR DESCRIPTION
**Pull request by man-shu**

### CODE REVIEW

File: ibc_public/utils_relaxo.py [hunk 1]
```diff
@@ -616,3 +619,107 @@ def t1_pipeline(do_normalise_before=False,
...
+        # preprocessing step: segmenting largest flip angle image
+        if do_segment:
+            return NotImplementedError
+    
+    # TO IMPLEMENT: extract echo times from NIFTIs or json sidecars
+    # mag_header = dict(mag_image.header)
+    # phase_header = dict(phase_image.header)
+
+    # hard-coding echo times for now
+    TEs = [2.03, 5.84, 9.65, 13.46, 17.27, 21.08, 24.89, 28.7, 32.51, 36.32, 40.13, 43.94]
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:711]: Would be better to have TEs in a side (json) file

Comment (man-shu) [reply ibc_public/utils_relaxo.py:711]: Yes, I am trying to find a way to get that info in json files after dcm2nii conversion step. It seems to get lost if 4d nifti images are created. So I've hardcoded it for now.

Comment (bthirion) [reply ibc_public/utils_relaxo.py:711]: My point is: I'd rather have it in a manually created json file than hard-coded here.
Thx.

File: ibc_public/utils_relaxo.py [hunk 2]
```diff
@@ -616,3 +620,200 @@ def t1_pipeline(do_normalise_before=False,
...
+
+        # preprocessing step: segmenting for masking
+        if do_segment:
+            time_elapsed = time.time() - start_time
+            # if already normalised, use MNI mask
+            if do_normalise_before:
+                print('[INFO,  t={:.2f}s] creating a mask'.format(time_elapsed))
+                image = norm_mag_nifti
+                mni = compute_brain_mask(image)
+                closed_mni = closing(mni, 12)
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:726]:
```suggestion
                closed_mni = closing(mni, closing_param)
```
where `closing_param = 12`

File: ibc_public/utils_relaxo.py [hunk 3]
```diff
@@ -616,3 +620,200 @@ def t1_pipeline(do_normalise_before=False,
...
+    # TO DO: extract echo times from NIFTIs or json sidecars
+    # mag_header = dict(mag_image.header)
+    # phase_header = dict(phase_image.header)
+
+    # import echo times
+    TE_file = open(echo_times)
+    TE_dict = json.load(TE_file)
+    TEs = TE_dict['TEs']
+
+    r2star_map, relative_uncertainty_map, aff, R2STARPath = monoexponential(image, TEs, len(TEs), recon_dir)
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:794]: line too long

**Comment by bthirion** [comment]
Please LMK when it's ready for review.

**Comment by man-shu** [comment]
Hi @bthirion, this branch is ready for review.
Thanks!

File: scripts/tmp_remove_files.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import os
```
Comment (bthirion) [inline scripts/tmp_remove_files.py:1]: Why are you removing these files ?

File: ibc_public/utils_relaxo.py [hunk 5]
```diff
@@ -1,50 +1,81 @@
...
+        path for the t2 image
+    t1_img : str
+        path for the t1 image
+    output_dir : str
+        location to write the corregistered image
+
+    Returns
+    -------
+    dict
+        dictionary object consisting of all input and output info including the
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:75]: Then you probably want to state what are the keys of the dictionary.

File: ibc_public/utils_relaxo.py [hunk 6]
```diff
@@ -56,10 +87,28 @@ def to_T2space(t2_img, t1_img, output_dir):
...
+        dict-like object containing information about preprocessing, input and
+        output file locations, by default SubjectData()
+    func : str, optional
+        path to another low-resolution(?) image to be normalised alongwith the
+        anatomical image. Used in t1_pipeline to normalise B1 map as functional
+        image with T1 map as anatomical, by default None
+
+    Returns
+    -------
+    dict
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:107]: Here too

File: ibc_public/utils_relaxo.py [hunk 7]
```diff
@@ -72,23 +121,52 @@ def to_MNI(image, data = SubjectData(), func=None):
...
+    image : str
+        path for the nifti image
+    normalize : bool
+        whether or not to normalize the segments
+
+    Returns
+    -------
+    dict
+        dictionary object consisting of all input and output info including the
+        name of output file which is used later in the pipelines
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:138]: +1

File: ibc_public/utils_relaxo.py [hunk 8]
```diff
@@ -380,31 +515,67 @@ def t2_pipeline(do_coreg=True, do_normalise_before=False,
...
+    do_segment : bool, optional
+        whether or not to segment the image. The segments are used to create a
+        mask used to exclude skull and neck during estimation. So if False,
+        estimation is done on every voxel in the input image, by default True
+    do_normalise_after : bool, optional
+        whether or not to normalise AFTER estimation. So if True the estimated
+        map is transformed to MNI152 space. Make it True only if 
+        do_normalise_before is False and vice versa. But normalising the 
+        estimated map has been observed to give spurious results for IBC
+        subjects, by default False
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:559]: You probably want to enforce that only one of `do_normalise_before` and `do_normalise_after` can be true.

File: ibc_public/utils_relaxo.py [hunk 9]
```diff
@@ -607,12 +785,265 @@ def t1_pipeline(do_normalise_before=False,
...
```
Comment (bthirion) [inline ibc_public/utils_relaxo.py:828]: Same comment on `do_normalise_before` and `do_normalise_after`.

**Review by bthirion — COMMENTED** [review]
That looks mostly good, simply a few minor things can be improved.

File: scripts/tmp_remove_files.py (hunk 4 above)
Comment (man-shu) [reply scripts/tmp_remove_files.py:1]: This was just a temporary script that I was using to remove previous runs, unnecessary files on margaret, my local machine etc. I plan on removing the script before merging

File: ibc_public/utils_relaxo.py (hunk 5 above)
Comment (man-shu) [reply ibc_public/utils_relaxo.py:75]: realised that this is just the `SubjectData()` object defined by pypreprocess, updating that

File: ibc_public/utils_relaxo.py (hunk 8 above)
Comment (man-shu) [reply ibc_public/utils_relaxo.py:562]: added assertion to do so

**Comment by man-shu** [comment]
Hi @bthirion! I think I have resolved the issues you mentioned. Could you please have another look?
Thanks!

**Review by bthirion — APPROVED** [review]
LGTM.

**Comment by bthirion** [comment]
Merging, then.

**Pull request merged**